server/uno_test.go

```diff
@@ -18,11 +19,367 @@ func setupGameWithPlayer(database *db.DB) (*model.Game, *model.Player) {
+	game.DiscardPile = []model.Card{model.Card{Color: "red", Value: "0"}}
+	game.Direction = true
+
+	game.Players[0].Cards = []model.Card{
+		model.Card{Color: "red", Value: "2"},
+		model.Card{Color: "red", Value: "R"},
+		model.Card{Color: "red", Value: "D2"},
+		model.Card{Color: "blue", Value: "6"},
+		model.Card{Color: "blue", Value: "7"},
+		model.Card{Color: "blue", Value: "9"},
+		model.Card{Color: "yellow", Value: "8"},
+	}
+
+	game.Players[1].Cards = []model.Card{
+		model.Card{Color: "red", Value: "S"},
+		model.Card{Color: "red", Value: "3"},
+		model.Card{Color: "green", Value: "6"},
+		model.Card{Color: "green", Value: "4"},
+		model.Card{Color: "blue", Value: "3"},
+		model.Card{Color: "blue", Value: "0"},
+		model.Card{Color: "blue", Value: "8"},
+	}
+
+	game.Players[2].Cards = []model.Card{
+		model.Card{Color: "red", Value: "D2"},
+		model.Card{Color: "red", Value: "2"},
+		model.Card{Color: "blue", Value: "9"},
+		model.Card{Color: "blue", Value: "5"},
+		model.Card{Color: "Black", Value: "W4"},
+		model.Card{Color: "blue", Value: "7"},
+		model.Card{Color: "green", Value: "8"},
+	}
+
+	game.Players[3].Cards = []model.Card{
+		model.Card{Color: "black", Value: "W"},
+		model.Card{Color: "red", Value: "3"},
+		model.Card{Color: "red", Value: "D2"},
+		model.Card{Color: "green", Value: "5"},
+		model.Card{Color: "blue", Value: "6"},
+		model.Card{Color: "blue", Value: "7"},
+		model.Card{Color: "yellow", Value: "8"},
+	}
+
+	database.SaveGame(*game)
+	return game
+}
+
+func TestPlayCard(t *testing.T) {
+
+	database, err := db.GetDb()
+
+	// Starting by passing in bogus information to make sure we get proper errors.
+	// Bogus game.
+	game, err := playCard("Bogus Game ID", "Bogus Player ID", model.Card{Color: "Bogus Color", Value: "Bogus Value"})
+
+	//Assert that the error is not nil, that it was returned
+	// Assert we didn't get a real game back from the bogus id
+	assert.NotNil(t, err)
+	assert.Nil(t, game)
+
+	game = setUpTestPlayerCard()
+
+	gameID := game.ID
+
+	//Real game, bogus player
+	game, err = playCard(game.ID, "Bogus Player ID", model.Card{Color: "Bogus Color", Value: "Bogus Value"})
+
+	// Assert there was an error, that the error is what we expect,
+	// Assert we didn't get a game back.
+	assert.NotNil(t, err)
+	assert.Equal(t, "You cannot participate in a game you do not belong", err.Error())
+	assert.Nil(t, game)
+
+	//Pull game from database to do checks that it wasn't affected.
+	game, _ = database.LookupGameByID(gameID)
+
+	// Assert that the discard pile does not have a new card on it.
+	// Assert that the two players still have all their cards.
+	// Assert that the player turn did not change
+	assert.Equal(t, 1, len(game.DiscardPile))
+	assert.Equal(t, 7, len(game.Players[0].Cards))
+	assert.Equal(t, 7, len(game.Players[1].Cards))
+	assert.Equal(t, 7, len(game.Players[2].Cards))
+	assert.Equal(t, 7, len(game.Players[3].Cards))
+	assert.Equal(t, 0, game.CurrentPlayer)
+
+	game = setUpTestPlayerCard()
```

Comment:
So. in unit testing... it's absolutely appropriate and more recommended to have each test only test one thing.

In go, that woujld probably look like:

```
TestPlayCard_DiscardSize(t *testing.T) {
    game := setUpTestPlayerCard()
    ...
}

TestPlayCard_HandSize(t *testing.T) {
    game := setUpTestPlayerCard()
    ...
}
... 
etc
```

I would recommend breaking up this larger test into smaller segments of tests, and having each test only focus on 1 aspect of the game. 

BTW, the formatting of the unit test name above is a set convention for Go. One of the recommendations are  `TestPlayCard_${TEST_DESCRIPTION}`. The only important part is the first section `TestPlayCard_`. Whatever comes after is up to you.

Reply:
Sure thing. I can make those changes.

Reply:
I have split the giant test into separate tests.